src/sas/sascalc/fit/BumpsFitting.py

```diff
@@ -282,77 +282,124 @@ def fit(self, msg_q=None,
-                'covariance']
-            assert values is not None
-            # Turn all parameters (fixed and varying) into uncertainty objects with
-            # zero uncertainty.
-            for p in unique_parameters(problem.model_parameters()):
-                p.value = ufloat(p.value, 0)
-            # then update the computed standard deviation of fitted parameters
-            if len(varying) < 2:
-                varying[0].value = ufloat(values[0], errs[0])
-            else:
-                fitted = (correlated_values(values, cov) if not errs else
-                          [ufloat(value, np.nan) for value in values])
 
-                for p, v in zip(varying, fitted):
-                    p.value = v
+        # TODO: shouldn't reference internal parameters of fit problem
+        varying = problem._parameters
+
+        values, errs, cov = result['value'], result['stderr'], result[
+            'covariance']
+        assert values is not None and errs is not None
+
+        # Propagate uncertainty through the parameter expressions
+        # We are going to abuse bumps a little here and stuff uncertainty
+        # objects into the parameter values, then update all the
+        # derived parameters with uncertainty propagation. We need to
+        # avoid triggering a model recalc since the uncertainty objects
+        # will not be working with sasmodels
+
+        if len(varying) < 2:
+            # Use the standard error as the error in the parameter
+            for param, val, err in zip(varying, values, errs):
+                # Convert all varying parameters to uncertainties objects
+                param.value = uncertainties.ufloat(val, err)
+        else:
             try:
-                problem.setp_hook()
-            except np.linalg.LinAlgError:
-                fitted = [ufloat(value, err) for value, err in zip(
-                    values, errs)]
-                uncertainty_error = True
-                for p, v in zip(varying, fitted):
-                    p.value = v
-                problem.setp_hook()
-            except TypeError:
-                propagate_errors = False
-                uncertainty_error = True
+                uncertainties.correlated_values(values, cov)
+            except:
+                # No convergance 
+                for param, val, err in zip(varying, values, errs):
+                    # Convert all varying parameters to uncertainties objects
+                    param.value = uncertainties.ufloat(val, err)
+            else:
+                # Use the covariance matrix to calculate error in the parameter
+                fitted = uncertainties.correlated_values(values, cov)
+                for param, val in zip(varying, fitted):
+                    param.value = val
+
+        # Propagate correlated uncertainty through constraints.
+        problem.setp_hook()
 
         # collect the results
         all_results = []
+
```

Comment:
Stylistic issue (and not introduced by @Caddy-Jones) but I think we try to avoid one letter variables (M, R).